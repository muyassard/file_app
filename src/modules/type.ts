export interface Image {
  photo: string;
  status: any;
  file: File | null;
  id: number;
}
