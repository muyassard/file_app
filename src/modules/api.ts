const baseUrl = "http://192.168.100.241:9999";
import axios from "axios";

export const CreateImage = (data: File) => {
  const formData = new FormData();
  formData.append("tenantId", "test");
  formData.append("module", "test");
  formData.append("fileName", data.name);
  formData.append("file", data);
  return axios.post(`${baseUrl}/api/file/upload/public`, formData);
};
