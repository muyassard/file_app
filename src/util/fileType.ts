

export const fileType = {
    pdf: "https://play-lh.googleusercontent.com/9XKD5S7rwQ6FiPXSyp9SzLXfIue88ntf9sJ9K250IuHTL7pmn2-ZB0sngAX4A2Bw4w",
    zip: "https://d1nhio0ox7pgb.cloudfront.net/_img/g_collection_png/standard/512x512/folder_zip.png",
    sql: "https://www.shareicon.net/data/2015/09/07/97430_document_512x512.png",
    html: "https://cdn4.iconfinder.com/data/icons/smashicons-file-types-flat/56/22_-_HTML_File_Flat-512.png",
    unknown: 'https://www.iconpacks.net/icons/2/free-file-icon-1453-thumb.png'
}


export const getFileType = (item: File) => {
    if (item.type.includes('png') || item.type.includes('jpg') || item.type.includes('jpeg')) return URL.createObjectURL(item)
    else if (item.type.includes('pdf')) return fileType.pdf;
    else if (item.type.includes('zip')) return fileType.zip;
    else if (item.type.includes('sql')) return fileType.sql;
    else if (item.type.includes('html')) return fileType.html;
    else { return fileType.unknown }
}