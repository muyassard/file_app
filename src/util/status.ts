import check from "@/images/check.png";
import error from "@/images/error.png";
import fail from "@/images/fail.png";


export const Status = {
    success: check,
    failed: fail,
    error: error,
}