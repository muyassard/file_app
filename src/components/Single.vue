<script setup lang="ts">
  import { CreateImage } from "@/modules/api";

  import defaultImage from "@/images/defaultImage.png";

  import { getFileType } from "@/util/fileType";
  import type { Image } from "@/modules/type";

  import { ref } from "vue";

  import { message } from "antd";
  import { Status } from "@/util/status";

  interface Props {
    maxSize: number;
    type: string;
  }

  const props = defineProps<Props>();

  const image = ref<Image>({
    photo: "",
    status: "",
    file: null,
    id: 0,
  });

  const isLoading = ref(false);
  const errorMessage = ref(false);

  const singleChange = async (e: Event) => {
    image.value.status = "";

    const files = (e.target as HTMLInputElement).files;
    const selectedFile = files?.[0];
    if (selectedFile) {
      if (!selectedFile.type.includes(props.type)) {
        errorMessage.value = true;
        message.error("you cannot upload such a file");
      } else if (selectedFile.size > props.maxSize) {
        errorMessage.value = true;
        image.value.status = Status.failed;
        message.error("This image size is large");
      } else {
        errorMessage.value = false;
        image.value.photo = getFileType(selectedFile);
        image.value.file = selectedFile;
        image.value.id = selectedFile.lastModified;
        console.log(image.value.file);
      }
    }
  };

  const uploadSingle = async (data: Image) => {
    if (data.file) {
      if (data.status) {
        message.error("you have uploaded a file");
        return;
      } else {
        try {
          errorMessage.value = false;
          isLoading.value = true;
          const imageUrl = await CreateImage(data.file);
          console.log(imageUrl);
          data.status = Status.success;
        } catch (error) {
          errorMessage.value = true;

          data.status = Status.error;
        } finally {
          isLoading.value = false;
        }
      }
    } else {
      message.error("you have not selected a file yet");
    }
  };
  defineExpose({
    singleChange,
    uploadSingle,
  });
</script>

<template>
  <div class="grid gap-2 w-[200px]">
    <div>Single</div>
    <label
      :style="{
        backgroundImage: `url(${image.photo || defaultImage})`,
      }"
      class="image border relative"
    >
      <img class="absolute top-0 left-1 w-[30px]" :src="image.status" />
      <input @change="singleChange" type="file" class="hidden" />
    </label>

    <div class="">{{ image?.file?.name }}</div>
    <p error-message v-if="errorMessage" class="">error</p>
    <b class="">you can only upload a {{ props.type }} file</b>

    <button
      data-type="singleUploadButton"
      @click="uploadSingle(image)"
      :disabled="isLoading"
      class="text-white bg-blue-500 w-[200px]"
    >
      {{ isLoading ? "uploading..." : "upload" }}
    </button>
  </div>
</template>

<style scoped>
  .image {
    width: 200px;
    height: 200px;
    background-position: center;
    background-size: cover;
  }
</style>
