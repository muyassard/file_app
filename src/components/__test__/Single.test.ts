import { CreateImage } from './../../modules/api';
import { describe, expect, it } from 'vitest'
import { mount } from '@vue/test-utils';

import Single from '../Single.vue';
import { fileType, getFileType } from '@/util/fileType';
import { nextTick } from 'vue';


const max = 1024 * 1024 * 4;


describe('Single', () => {

    it("renders a file input", () => {
        const wrapper = mount(Single);
        expect(wrapper.exists()).toBe(true)
        expect(wrapper.find("input").html()).matchSnapshot()
    });

    it('renders file type based on file type correctly', async () => {
        const file = new File(["1200"], "fake.pdf", { type: "application/pdf" });
        const type = getFileType(file)
        expect(type).toBe(fileType.pdf)
    });

    it('displays an error message when selecting the wrong file type', async () => {
        const wrapper = mount(Single, {
            props: {
                maxSize: 2000000,
                type: 'zip',
            },
        });

        const inputElement = wrapper.find('input').element as HTMLInputElement;
        const file = new File(['2777'], 'foo.pdf', {
            type: 'application/pdf',
        });
        const mockFileList = Object.create(inputElement.files);
        mockFileList[0] = file;
        Object.defineProperty(mockFileList, 'length', { value: 1 });
        (wrapper.getCurrentComponent().exposed as unknown as any).singleChange({
            target: { files: mockFileList },
        });



        await nextTick();

        const errorMessage = wrapper.find('[error-message]')

        expect(errorMessage.exists()).toBe(true);
    });

    it('shows an error message when selecting a large file', async () => {
        const wrapper = mount(Single, {
            props: {
                maxSize: 2,
                type: 'pdf',
            },
        });

        const inputElement = wrapper.find('input').element as HTMLInputElement;
        const file = new File(['2777'], 'foo.pdf', {
            type: 'application/pdf',
        });
        const mockFileList = Object.create(inputElement.files);
        mockFileList[0] = file;
        Object.defineProperty(mockFileList, 'length', { value: 1 });
        (wrapper.getCurrentComponent().exposed as unknown as any).singleChange({
            target: { files: mockFileList },
        });



        await nextTick();

        const errorMessage = wrapper.find('[error-message]')

        expect(errorMessage.exists()).toBe(true);
    });

    // it('does not show an error message when loading the correct file type', async () => {
    //     const wrapper = mount(Single, {
    //         props: {
    //             maxSize: 4,
    //             type: 'zip',
    //         },
    //     });

    //     const input = wrapper.find('input');


    //     await input.trigger('change', {
    //         files: [{ type: 'application/zip', size:  }],
    //     });

    //     await nextTick();

    //     expect(wrapper.find('p[error-message]').exists()).toBe(false);
    // });


    it('api ', async () => {
        const wrapper = mount(Single, {
            props: {
                maxSize: 1000000,
                type: 'pdf',
            },
        });
        const data = 'Hello world';

        const file = new File([data], "fake.pdf", { type: "application/pdf" });

        await CreateImage(file);

        await nextTick();

        expect(wrapper.find('[error-message]').exists()).toBe(false);
    });

});














