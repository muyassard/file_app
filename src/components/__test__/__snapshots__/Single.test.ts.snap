// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Single > renders a file input 1`] = `"<input data-v-91122bb6="" type="file" class="hidden">"`;

exports[`Single > should render File input 1`] = `"<input data-v-91122bb6="" type="file" class="hidden">"`;

exports[`Single > should return correctly 1`] = `"<input data-v-91122bb6="" data-type="singleInput" type="file" class="hidden">"`;

exports[`change > should return correctly 1`] = `
"<div data-v-91122bb6="" class="grid gap-2 w-[200px]">
  <div data-v-91122bb6="">Single</div><label data-v-91122bb6="" style="background-image: url(/src/images/defaultImage.png);" class="image border relative"><img data-v-91122bb6="" class="absolute top-0 left-1 w-[30px]" src=""><input data-v-91122bb6="" type="file" class="hidden"></label>
  <div data-v-91122bb6="" class=""></div><b data-v-91122bb6="" class="">you can only upload a file</b><button data-v-91122bb6="" class="text-white bg-blue-500 w-[200px]">upload</button>
</div>"
`;

exports[`sum > should return correctly 1`] = `"0"`;
